### PR TITLE
Sample plant location feature

### PR DESCRIPTION
**Pull request by sunilsabatp**
This PR handles sample plant location feature.

Resolves minor bug in loading project list on back (https://github.com/Plant-for-the-Planet-org/planet-webapp/pull/2209#discussion_r1768406837)

### CODE REVIEW

File: public/static/locales/en/projectDetails.json [hunk 1]
```diff
@@ -25,9 +25,17 @@
     "plantingDensityUnit": "{formattedCount} trees per ha",
     "speciesCount": "Species Planted ({count})",
     "sampleTrees": "Sample Trees ({count})",
+    "sampleTree": "Sample Tree",
+    "measurment": "Measurment",
     "speciesMeasurment": "Tag #{hid} • {plantHeight}m high • {plantWidth}cm wide",
+    "singleSpeciesMeasurment": "{plantHeight}m high • {plantWidth}cm wide",
```
Comment (mohitb35) [inline public/static/locales/en/projectDetails.json:31]: Fix the typos in the text/keys

It should be "measurement"

File: src/features/projectsV2/ProjectDetails/utils.tsx [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+    } else if (typeof data === 'object' && data !== null) {
+      if (data.image) {
+        Images.push({ image: data.image });
+      }
+      Object.values(data).forEach(traverse);
+    }
+  }
+  traverse(sampleInterventionsImages);
+  return Images;
+};
```
Comment (mohitb35) [inline src/features/projectsV2/ProjectDetails/utils.tsx:22]: This seems to be unused in the project

File: src/features/projectsV2/ProjectsMap/index.tsx [hunk 3]
```diff
@@ -48,11 +49,17 @@ function ProjectsMap(props: ProjectsMapProps) {
...
+        setHoveredPl(null);
+      }
     },
-    [plantLocations]
+    [plantLocations, selectedPl]
   );
-
   const onClick = useCallback(
     (e) => {
       const selectedPlantLocation = getPlantLocationInfo(
```
Comment (mohitb35) [inline src/features/projectsV2/ProjectsMap/index.tsx:65]: Currently, I cannot unset a sample plant location once selected, if I click on the containing plant location. The only way to do so is by clicking on a different plant location and then coming back to the first one.

File: src/features/projectsV2/ProjectDetails/components/SamplePlantInfo.tsx [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+import { SamplePlantLocation } from '../../../common/types/plantLocation';
+import getImageUrl from '../../../../utils/getImageURL';
+
+interface Props {
+  samplePlantData: SamplePlantLocation;
+}
+
+const SamplePlantInfo = ({ samplePlantData }: Props) => {
+  const t = useTranslations('ProjectDetails');
+  const { hid, plantDate, tag, scientificName, measurements } = samplePlantData;
```
Comment (mohitb35) [inline src/features/projectsV2/ProjectDetails/components/SamplePlantInfo.tsx:14]: Use interventionStartDate instead of plantDate

File: src/features/projectsV2/ProjectDetails/components/microComponents/SamplePlantInfoCard.tsx [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+              plantWidth: measurements.width,
+            })}
+          </p>
+        </div>
+      )}
+      <div className={styles.plantingDetailsSubContainer}>
+        <p className={styles.label}>{t('plot')}</p>
+        <button className={styles.showWholeArea}>{t('showWholeArea')}</button>
+      </div>
+    </>
```
Comment (mohitb35) [inline src/features/projectsV2/ProjectDetails/components/microComponents/SamplePlantInfoCard.tsx:67]: labels should be a heading element

File: src/features/projectsV2/ProjectDetails/components/microComponents/SamplePlantInfoCard.tsx [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+              plantHeight: measurements.height,
+              plantWidth: measurements.width,
+            })}
+          </p>
+        </div>
+      )}
+      <div className={styles.plantingDetailsSubContainer}>
+        <p className={styles.label}>{t('plot')}</p>
+        <button className={styles.showWholeArea}>{t('showWholeArea')}</button>
+      </div>
```
Comment (mohitb35) [inline src/features/projectsV2/ProjectDetails/components/microComponents/SamplePlantInfoCard.tsx:66]: I don't think this is ready yet. What action is supposed to happen here.

File: src/features/projectsV2/ProjectDetails/components/ImageCarousel.tsx [hunk 7]
```diff
@@ -3,9 +3,10 @@ import Stories from 'react-insta-stories';
 import getImageUrl from '../../../../utils/getImageURL';
 import { SliderImage } from '../../../projects/components/PlantLocation/ImageSlider';
 import { SingleCarouselImage } from './microComponents/SingleCarouselImage';
+import { PlantLocationCoordinate } from '../../../common/types/plantLocation';
 
 interface Props {
-  images: SliderImage[];
+  images: SliderImage[] | PlantLocationCoordinate[] | undefined;
```
Comment (mohitb35) [inline src/features/projectsV2/ProjectDetails/components/ImageCarousel.tsx:9]: Why was the type updated?

File: src/features/projectsV2/ProjectDetails/components/PlantLocationInfo.tsx [hunk 8]
```diff
@@ -49,8 +50,8 @@ const PlantLocationInfoSection = ({
     }
   }, [isMultiTreeRegistration ? plantLocationInfo.sampleInterventions : null]);
 
-  const hasSampleInterventionSpeciesImages =
-    sampleInterventionSpeciesImages !== undefined &&
+  const shouldDisplayImageCarousel =
+    sampleInterventionSpeciesImages &&
     sampleInterventionSpeciesImages?.length > 0;
```
Comment (mohitb35) [inline src/features/projectsV2/ProjectDetails/components/PlantLocationInfo.tsx:55]: This should be a boolean value, not boolean or undefined

File: src/features/projectsV2/ProjectDetails/components/PlantLocationInfo.tsx [hunk 9]
```diff
@@ -13,11 +13,12 @@ import TreeMapperBrand from './microComponents/TreeMapperBrand';
 import PlantingDetails from './microComponents/PlantingDetails';
 import { useTranslations } from 'next-intl';
 
-const PlantLocationInfoSection = ({
+const PlantLocationInfo = ({
   plantLocationInfo,
 }: {
   plantLocationInfo: PlantLocation | SamplePlantLocation | null;
 }) => {
+  const t = useTranslations('ProjectDetails');
```
Comment (mohitb35) [inline src/features/projectsV2/ProjectDetails/components/PlantLocationInfo.tsx:21]: We don't seem to need the translation function in this component

File: src/features/projectsV2/ProjectDetails/components/microComponents/SamplePlantInfoCard.tsx [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+            </div>
+          ) : null;
+        })}
+      </div>
+      {scientificName && (
+        <div className={styles.plantingDetailsSubContainer}>
+          <p className={styles.label}>{t('scientificName')}</p>
+          <p className={styles.data}>{scientificName}</p>
+        </div>
+      )}
```
Comment (mohitb35) [inline src/features/projectsV2/ProjectDetails/components/microComponents/SamplePlantInfoCard.tsx:51]: Using `plantingDetailsSubContainer` when there is no outside container is a little confusing. Perhaps rename `plantingDetailsContainer` >> `plantingDetailsGroup` and `plantingDetailsSubContainer` to `plantingDetailsItem`?

**Review by mohitb35 — CHANGES_REQUESTED** [review]
@sunilsbatp Please go through my feedback. Nothing major.

File: src/features/projectsV2/ProjectsMap/utils.tsx [hunk 11]
```diff
@@ -32,7 +32,7 @@ export const getPlantLocationInfo = (
   }
   const map = mapRef.current.getMap();
   const features = map.queryRenderedFeatures(point, {
-    layers: ['plant-polygon-layer'],
+    layers: ['plant-polygon-layer', 'point-layer'],
   });
```
[the remark this answers is not in the document]
Comment (sunilsabatp) [reply src/features/projectsV2/ProjectsMap/utils.tsx]: point-layer  has been removed

File: src/features/projectsV2/ProjectDetails/components/microComponents/SamplePlantInfoCard.tsx (hunk 6 above)
Comment (sunilsabatp) [reply src/features/projectsV2/ProjectDetails/components/microComponents/SamplePlantInfoCard.tsx:66]: I'm not fully sure about this; should we hide this portion?

File: src/features/projectsV2/ProjectDetails/components/ImageCarousel.tsx (hunk 7 above)
Comment (sunilsabatp) [reply src/features/projectsV2/ProjectDetails/components/ImageCarousel.tsx:9]: PlantLocationCoordinate[] has been removed

**Comment by mohitb35** [comment]
Should the Powered by Treemapper logo be seen for sample plant locations as well?

**Comment by sunilsabatp** [comment]
> Should the Powered by Treemapper logo be seen for sample plant locations as well?



> Should the Powered by Treemapper logo be seen for sample plant locations as well?

Yes, we should display it because it is present in the Figma design as well as in our development environment (dev.pp.eco)

File: src/features/projectsV2/ProjectsMap/utils.tsx (hunk 11 above)
[the remark this answers is not in the document]
Comment (mohitb35) [reply src/features/projectsV2/ProjectsMap/utils.tsx]: Plant locations can be "single-tree-registrations" as well, which would have a Point geometry. "multi-tree-registrations" can also have a point geometry.

e.g. for the project `green4life`

File: src/features/projectsV2/ProjectsMap/index.tsx [hunk 12]
```diff
@@ -91,9 +91,7 @@ function ProjectsMap(props: ProjectsMapProps) {
         attributionControl={false}
         ref={mapRef}
         interactiveLayerIds={
-          shouldShowSingleProjectsView
-            ? ['plant-polygon-layer', 'point-layer']
-            : undefined
+          shouldShowSingleProjectsView ? ['plant-polygon-layer'] : undefined
```
Comment (mohitb35) [inline src/features/projectsV2/ProjectsMap/index.tsx:94]: How does a user click on a
1. A `single-tree-registration`
2. A `multi-tree-registration` with point geometry

File: src/features/projectsV2/ProjectsMap/microComponents/PlantLocations.tsx [hunk 13]
```diff
@@ -34,6 +34,10 @@ export default function PlantLocations(): ReactElement {
   const locale = useLocale();
 
   const openPl = (pl: PlantLocationSingle | SamplePlantLocation) => {
+    if (selectedSamplePlantLocation?.hid === pl.hid) {
```
Comment (mohitb35) [inline src/features/projectsV2/ProjectsMap/microComponents/PlantLocations.tsx:37]: This allows the user to unselect the sample plant location while clicking on it again. However, I feel the sample plant location should also get unselected if the user clicks on the surrounding plant location again (this does not currently happen)

File: src/features/projectsV2/ProjectsMap/microComponents/PlantLocations.tsx
Comment (mohitb35) [inline src/features/projectsV2/ProjectsMap/microComponents/PlantLocations.tsx:1]: Another thing to consider - what is supposed to happen when I hover above a sample plant location (point)? We intend to call `onHover`, but nothing actually seems to happen.

File: src/features/projectsV2/ProjectDetails/components/microComponents/SamplePlantInfoCard.tsx (hunk 6 above)
Comment (mohitb35) [reply src/features/projectsV2/ProjectDetails/components/microComponents/SamplePlantInfoCard.tsx]: Have asked for clarification.

File: src/features/projectsV2/ProjectsMap/microComponents/PlantLocations.tsx
Comment (sunilsabatp) [reply src/features/projectsV2/ProjectsMap/microComponents/PlantLocations.tsx:1]: Since we are not showing the hover value for the sample plant location in the sidebar, I have removed the hover functionality logic for the sample plant location

**Review by mariahosfeld — CHANGES_REQUESTED** [review]
Please check the following points:
- [x] When selecting an older location, that is on a deeper layer than a most recent one, I am able to see the sample locations on the map, but cannot select them that way. Always the location "above" it will be selected instead (tested with /prd/yucatan)
- [x] After selecting a sample location, clicking on its parent does not change the displayed information, only hovering or clicking a different sample location or plant location will do that (tested with /prd/yucatan)
- [x] After selecting a plant location, the drop-down for the sites does no longer show the name of the site selected, is this intended? (tested with /prd/yucatan)
image
- [x] for the mentioned example of /prd/green4life I can only select the multi location, but not the single one
- [x] for the mentioned example of /prd/mothersforest I only see a location marker, nothing else
image

**Review by mariahosfeld — COMMENTED** [review]
Issues listed yesterday now resolved.

**Review by mariahosfeld — APPROVED** [review]

File: src/features/projectsV2/ProjectsMap/utils.tsx (hunk 11 above)
[the remark this answers is not in the document]
Comment (sunilsabatp) [reply src/features/projectsV2/ProjectsMap/utils.tsx]: will cover in a separate PR

File: src/features/projectsV2/ProjectDetails/components/microComponents/SamplePlantInfoCard.tsx (hunk 6 above)
Comment (sunilsabatp) [reply src/features/projectsV2/ProjectDetails/components/microComponents/SamplePlantInfoCard.tsx]: will cover in a separate PR

Comment (mohitb35) [reply src/features/projectsV2/ProjectDetails/components/microComponents/SamplePlantInfoCard.tsx]: added in https://github.com/Plant-for-the-Planet-org/planet-webapp/pull/2183

**Review by mariahosfeld — COMMENTED** [review]
Amended time for zoom into plant location looks good.

File: src/features/projectsV2/ProjectDetails/components/SinglePlantInfo.tsx [hunk 14]
```diff
@@ -0,0 +1,48 @@
...
+import { useTranslations } from 'next-intl';
+import {
+  PlantLocation,
+  SamplePlantLocation,
+} from '../../../common/types/plantLocation';
+import getImageUrl from '../../../../utils/getImageURL';
+import TreeMapperBrand from './microComponents/TreeMapperBrand';
+
+interface Props {
+  plantData: SamplePlantLocation | PlantLocation | null;
```
Comment (mohitb35) [inline src/features/projectsV2/ProjectDetails/components/SinglePlantInfo.tsx:13]: Update types to make more specific. This does not allow multiple tree registrations.

File: src/features/projectsV2/ProjectDetails/index.tsx [hunk 15]
```diff
@@ -103,6 +106,43 @@ const ProjectDetails = ({
       loadPlantLocations();
   }, [singleProject]);
 
+  const shouldShowPlantLocationInfo =
+    hoveredPlantLocation?.type === 'multi-tree-registration' ||
+    selectedPlantLocation?.type === 'multi-tree-registration';
+  !isMobile;
```
Comment (mohitb35) [inline src/features/projectsV2/ProjectDetails/index.tsx:112]: There appears to be some extra code here

File: src/features/projectsV2/ProjectDetails/components/PlantLocationInfo.tsx [hunk 16]
```diff
@@ -5,19 +5,20 @@ import {
...
-  plantLocationInfo,
-}: {
+interface Props {
   plantLocationInfo: PlantLocation | SamplePlantLocation | null;
-}) => {
+  isMobile: boolean;
+}
+
+const PlantLocationInfo = ({ plantLocationInfo, isMobile }: Props) => {
   const isMultiTreeRegistration =
```
Comment (mohitb35) [inline src/features/projectsV2/ProjectDetails/components/PlantLocationInfo.tsx:22]: Since this is only for multi-tree-registrations, updates types and logic accordingly. Also, name the component MultiPlantLocationInfo

File: src/features/projectsV2/ProjectDetails/components/SinglePlantInfo.tsx [hunk 17]
```diff
@@ -0,0 +1,48 @@
...
+}
+
+const SinglePlantInfo = ({ plantData }: Props) => {
+  const t = useTranslations('ProjectDetails');
```
Comment (mohitb35) [inline src/features/projectsV2/ProjectDetails/components/SinglePlantInfo.tsx:17]: Make a minor change to call this SinglePlantLocationInfo

File: src/features/projectsV2/ProjectDetails/components/microComponents/SingleCarouselImage.tsx [hunk 18]
```diff
@@ -1,44 +1,37 @@
-import { useMemo } from 'react';
 import { SliderImage } from '../../../../projects/components/PlantLocation/ImageSlider';
 import styles from '../../styles/Slider.module.scss';
```
Comment (mohitb35) [inline src/features/projectsV2/ProjectDetails/components/microComponents/SingleCarouselImage.tsx:2]: Avoid using code from the old version of projects. Set up the SliderImage type within v2 itself.

File: src/features/projectsV2/ProjectDetails/components/ImageCarousel.tsx [hunk 19]
```diff
@@ -1,11 +1,11 @@
-import React, { ReactElement, useEffect } from 'react';
+import React, { ReactElement, useEffect, useState } from 'react';
 import Stories from 'react-insta-stories';
 import getImageUrl from '../../../../utils/getImageURL';
 import { SliderImage } from '../../../projects/components/PlantLocation/ImageSlider';
 import { SingleCarouselImage } from './microComponents/SingleCarouselImage';
```
Comment (mohitb35) [inline src/features/projectsV2/ProjectDetails/components/ImageCarousel.tsx:5]: Avoid using code from the old version of projects. Set up the SliderImage type within v2 itself.

File: src/features/projectsV2/ProjectDetails/components/microComponents/ImageSlider.tsx [hunk 20]
```diff
@@ -10,8 +10,14 @@ interface ImageSliderProps {
   images: Image[];
```
Comment (mohitb35) [inline src/features/projectsV2/ProjectDetails/components/microComponents/ImageSlider.tsx:10]: Since we don't always pass in the images property from the projects API, it may be better to define the type for Image that is used in ImageSlider rather than refer to the common type defined. Makes the code easier to read.

File: src/features/projectsV2/ProjectsMap/SingleProjectView.tsx [hunk 21]
```diff
@@ -54,15 +63,52 @@ const SingleProjectView = ({ mapRef }: { mapRef: MapRef }) => {
...
+      if (lon !== undefined && lat !== undefined) {
+        // Zoom into single tree registration
+        zoomToLocation(setViewState, lon, lat, 20, 3500, mapRef);
+      }
+    }
+  }, [
+    singleProject.sites,
+    selectedPlantLocation,
+    router.isReady,
+    selectedPlantLocation,
```
Comment (mohitb35) [inline src/features/projectsV2/ProjectsMap/SingleProjectView.tsx:96]: The deps are repeated

File: src/features/projectsV2/ProjectsMap/SingleProjectView.tsx [hunk 22]
```diff
@@ -54,15 +63,52 @@ const SingleProjectView = ({ mapRef }: { mapRef: MapRef }) => {
...
+      zoomToLocation(setViewState, longitude, latitude, 10, 3500, mapRef);
+    } else if (
+      selectedPlantLocation?.geometry?.type === 'Point' &&
+      selectedPlantLocation?.geometry.coordinates.length === 2
+    ) {
+      const [lon, lat] = selectedPlantLocation.geometry.coordinates;
+      if (lon !== undefined && lat !== undefined) {
+        // Zoom into single tree registration
+        zoomToLocation(setViewState, lon, lat, 20, 3500, mapRef);
+      }
```
Comment (mohitb35) [inline src/features/projectsV2/ProjectsMap/SingleProjectView.tsx:90]: Would it be better to separate the code for handling plant location zoom? i.e. group the code that handles zoom into point and polygon plant locations together.

File: src/features/projectsV2/ProjectsMap/index.tsx [hunk 23]
```diff
@@ -26,13 +31,28 @@ export type ProjectsMapProps = ProjectsMapMobileProps | ProjectsMapDesktopProps;
...
+    locale: string,
+    projectSlug: string,
+    siteIndex: number
+  ) => {
+    if (singleProject?.sites?.length === 0) return;
+    setSelectedSite(siteIndex);
+    const siteId = singleProject?.sites?.[siteIndex].properties.id;
+    if (siteId) updateUrlWithSiteId(locale, projectSlug, siteId, router);
+  };
+
```
Comment (mohitb35) [inline src/features/projectsV2/ProjectsMap/index.tsx:55]: Why was this added here, when we already have similar code in the ProjectsProvider?

My concern is that we are keeping the logic that updates the url in too many places, which could cause bugs down the line.

**Review by mohitb35 — CHANGES_REQUESTED** [review]
@sunilsabatp Please review my feedback

File: src/features/projectsV2/ProjectsMap/SingleProjectView.tsx (hunk 22 above)
Comment (sunilsabatp) [reply src/features/projectsV2/ProjectsMap/SingleProjectView.tsx:90]: Have covered in this (https://github.com/Plant-for-the-Planet-org/planet-webapp/pull/2255) PR

File: src/features/projectsV2/ProjectsMap/SingleProjectView.tsx (hunk 21 above)
Comment (sunilsabatp) [reply src/features/projectsV2/ProjectsMap/SingleProjectView.tsx:96]: Have covered in this (https://github.com/Plant-for-the-Planet-org/planet-webapp/pull/2255) PR

File: src/features/projectsV2/ProjectDetails/components/SinglePlantInfo.tsx (hunk 17 above)
Comment (sunilsabatp) [reply src/features/projectsV2/ProjectDetails/components/SinglePlantInfo.tsx]: Since the sidebar only displays plant information and not location information, I believe that` 'SinglePlantInfo'` is the correct name for the component

File: src/features/common/types/projectv2.d.ts [hunk 24]
```diff
@@ -15,3 +17,19 @@ export type MapProjectProperties =
...
+
+export type SliderImage = {
+  image?: string | undefined;
+  description?: string | null;
+};
+
+export type PointPlantLocation =
+  | PlantLocationSingle
+  | SamplePlantLocation
+  | null;
```
Comment (mohitb35) [inline src/features/common/types/projectv2.d.ts:35]: 1. Adding null to the type is confusing. 
2. Personally, I prefer just using the constituent types directly, since this is only used in 2 places, and the type only has 2 parts (excluding the null).

File: src/utils/projectV2.ts [hunk 25]
```diff
@@ -113,3 +126,67 @@ export const getPlantLocationInfo = (
...
+  siteId: string,
+  router: NextRouter
+) => {
+  const updatedQueryParams = updateUrlWithParams(
+    router.asPath,
+    router.query,
+    siteId
+  );
+  pushWithShallow(locale, projectSlug, updatedQueryParams, router);
+};
```
Comment (mohitb35) [inline src/utils/projectV2.ts:180]: Since these are only used in `ProjectsProvider`, and also because they need `router`, they could be left in `ProjectsProvider` itself.

**Review by mohitb35 — APPROVED** [review]
@sunilsabatp Take a look at the new feedback, however this is good to merge after addressing those points.
